### PR TITLE
[SPARK-1507][YARN]specify num of cores for AM 

### PR DESCRIPTION
**Pull request by XuTingjun**
I add some configurations below.
spark.yarn.am.cores/SPARK_MASTER_CORES/SPARK_DRIVER_CORES for yarn-client mode;
spark.driver.cores for yarn-cluster mode.

### CODE REVIEW

**Comment by XuTingjun** [comment]
@andrewor14

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/ClientArguments.scala [hunk 1]
```diff
@@ -120,6 +121,13 @@ private[spark] class ClientArguments(args: Array[String], sparkConf: SparkConf)
           amMemory = value
           args = tail
 
+        case ("--master-cores" | "--driver-cores") :: IntParam(value) :: tail =>
```
Comment (sryza) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/ClientArguments.scala]: no need to add in a deprecated "--master-cores" property

File: yarn/src/main/scala/org/apache/spark/scheduler/cluster/YarnClientSchedulerBackend.scala [hunk 2]
```diff
@@ -70,6 +70,8 @@ private[spark] class YarnClientSchedulerBackend(
       List(
         ("--driver-memory", "SPARK_MASTER_MEMORY", "spark.master.memory"),
         ("--driver-memory", "SPARK_DRIVER_MEMORY", "spark.driver.memory"),
+        ("--driver-cores", "SPARK_MASTER_CORES", "spark.yarn.am.cores"),
```
Comment (sryza) [inline yarn/src/main/scala/org/apache/spark/scheduler/cluster/YarnClientSchedulerBackend.scala]: No need to add in a deprecated SPARK_MASTER_CORES env var.

**Comment by sryza** [comment]
@XuTingjun ah, that's correct.  Looking more closely, my confusion was stemming from some existing weirdness, which is that setting the "spark.driver.memory" property will set the application master memory even in client mode.  Also, that passing "--driver-memory" as part of the YARN code's ClientArguments (but not to `spark-submit`) will do this as well.

My opinion is that we should fix those, though I suppose it could be argued that it breaks backwards compatibility?

**Comment by XuTingjun** [comment]
@sryza, do you mean "spark.driver.memory" works in yarn-client and yarn-cluster mode, so we should use one configuration maybe named "spark.driver.cores" to set am cores?

**Comment by sryza** [comment]
I think the best thing would be to split spark.driver.memory into spark.driver.memory and spark.yarn.am.memory, and to have the latter only work for the yarn-client AM.

**Comment by XuTingjun** [comment]
Yearh, I agree with you. Later I will fix this. Thanks @sryza

**Comment by XuTingjun** [comment]
@sryza, I have splited spark.driver.memory into spark.driver.memory and spark.yarn.am.memory. Please have a look.

**Comment by WangTaoTheTonic** [comment]
Sorry but I already file a PR that splits spark.driver.memory in https://github.com/apache/spark/pull/3607.

Could you please check if anyone already did the same work first?

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/ClientArguments.scala [hunk 3]
```diff
@@ -43,8 +44,13 @@ private[spark] class ClientArguments(args: Array[String], sparkConf: SparkConf)
 
   // Additional memory to allocate to containers
   // For now, use driver's memory overhead as our AM container's memory overhead
-  val amMemoryOverhead = sparkConf.getInt("spark.yarn.driver.memoryOverhead",
-    math.max((MEMORY_OVERHEAD_FACTOR * amMemory).toInt, MEMORY_OVERHEAD_MIN))
+  val amMemoryOverhead = if (userClass != null) {
+    sparkConf.getInt("spark.yarn.driver.memoryOverhead",
+      math.max((MEMORY_OVERHEAD_FACTOR * amMemory).toInt, MEMORY_OVERHEAD_MIN))
+  } else {
```
Comment (andrewor14) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/ClientArguments.scala]: indent

**Comment by andrewor14** [comment]
Hi @XuTingjun there seems to be a certain degree of overlap with work in #3607. Also, my concern with this PR is that it conflates "driver" and "AM" in a few places. For instance, `--driver-cores` should never configure the number of cores used by the AM in client mode.

**Pull request closed without merging**